src/lib/autoDetect.ts

```diff
@@ -0,0 +1,106 @@
+import {
+  ExtensionContext,
+  workspace,
+  window,
+  ConfigurationTarget,
+  WorkspaceConfiguration,
+} from 'vscode';
+import { FileAssociations } from './types';
+const { extname } = require('path');
+import {
+  EXTENSION_CONFIG_NAME,
+  EXTENSION_CONFIG_KEYS,
+  HUBL_HTML_ID,
+  HUBL_CSS_ID,
+} from './constants';
+
+export const initializeHubLAutoDetect = (context: ExtensionContext) => {
+  if (
+    hasHubLAssociations() ||
+    checkGlobalState() ||
+    !!context.workspaceState.get('DO_NOT_USE_HUBL')
+  ) {
+    return;
+  }
+  const showHublDetectedMessage = () => {
+    const hublDetectedMessage = `It looks like this file contains HubL code. Would you like to use HubL for this project?`;
+    const hublDetectedYesButton = `Use HubL`;
+    const hublDetectedNoButton = 'No';
+    const hublDetectedNeverAgainButton = 'Never ask again';
+
+    window
+      .showInformationMessage(
+        hublDetectedMessage,
+        hublDetectedYesButton,
+        hublDetectedNoButton,
+        hublDetectedNeverAgainButton
+      )
+      .then((selection) => {
+        switch (selection) {
+          case hublDetectedYesButton: {
+            updateWorkspaceFileAssociation();
+          }
+          case hublDetectedNoButton: {
+            context.workspaceState.update('DO_NOT_USE_HUBL', true);
+          }
+          case hublDetectedNeverAgainButton: {
+            workspace
+              .getConfiguration(EXTENSION_CONFIG_NAME)
+              .update(
+                EXTENSION_CONFIG_KEYS.NEVER_USE_HUBL,
+                true,
+                ConfigurationTarget.Global
+              );
+          }
+          default: // User closed the dialogue
+        }
+        handleTextDocumentOpen.dispose();
+      });
+  };
+
+  const handleTextDocumentOpen = workspace.onDidOpenTextDocument(() => {
+    const textDocument = window.activeTextEditor?.document;
+    if (!textDocument) {
+      return;
+    }
+    const fileExt = extname(textDocument.fileName);
+    if (fileExt !== '.html' && fileExt !== '.css') {
+      return;
+    }
+
+    let n = 1;
+    const limit = textDocument.lineCount;
+    while (n < 50 && n <= limit) {
```

Comment:
Same bit here about calculating the min of limit and 50 first and then only checking one condition in the loop.